linalg/blockoperator.hpp

```diff
@@ -14,129 +14,50 @@
  *
  * If a block is not set, it is assumed to be a zero block.
  */
-class BlockOperator : public Operator
-{
-public:
-   //! Constructor for BlockOperators with the same block-structure for rows and
-   //! columns.
-   /**
-    *  offsets: offsets that mark the start of each row/column block (size
-    *  nRowBlocks+1).
-    */
-   BlockOperator(const Array<int> & offsets);
-   //! Constructor for general BlockOperators.
-   /**
-    *  row_offsets: offsets that mark the start of each row block (size
-    *  nRowBlocks+1).  col_offsets: offsets that mark the start of each column
-    *  block (size nColBlocks+1).
-    */
-   BlockOperator(const Array<int> & row_offsets, const Array<int> & col_offsets);
-
-   /// Copy assignment is not supported
-   BlockOperator &operator=(const BlockOperator &) = delete;
-
-   /// Move assignment is not supported
-   BlockOperator &operator=(BlockOperator &&) = delete;
-
-   //! Add block op in the block-entry (iblock, iblock).
-   /**
-    * iblock: The block will be inserted in location (iblock, iblock).
-    * op: the Operator to be inserted.
-    * c: optional scalar multiple for this block.
-    */
-   void SetDiagonalBlock(int iblock, const Operator *op, double c = 1.0);
-   //! Add a block op in the block-entry (iblock, jblock).
-   /**
-    * irow, icol: The block will be inserted in location (irow, icol).
-    * op: the Operator to be inserted.
-    * c: optional scalar multiple for this block.
-    */
-   void SetBlock(int iRow, int iCol, const Operator *op, double c = 1.0);
-
-   //! Return the number of row blocks
-   int NumRowBlocks() const { return nRowBlocks; }
-   //! Return the number of column blocks
-   int NumColBlocks() const { return nColBlocks; }
-
-   //! Check if block (i,j) is a zero block
-   int IsZeroBlock(int i, int j) const { return (op(i,j)==NULL) ? 1 : 0; }
-   //! Return a reference to block i,j
-   const Operator & GetBlock(int i, int j) const
-   { MFEM_VERIFY(op(i,j), ""); return *op(i,j); }
-   //! Return the coefficient for block i,j
-   double GetBlockCoef(int i, int j) const
-   { MFEM_VERIFY(op(i,j), ""); return coef(i,j); }
-   //! Set the coefficient for block i,j
-   void SetBlockCoef(int i, int j, double c)
-   { MFEM_VERIFY(op(i,j), ""); coef(i,j) = c; }
+typedef TBlockOperator<const Operator> BlockConstOperator;
 
-   //! Return the row offsets for block starts
-   Array<int> & RowOffsets() { return row_offsets; }
-   //! Read only access to the row offsets for block starts
-   const Array<int> & RowOffsets() const { return row_offsets; }
-   //! Return the columns offsets for block starts
-   Array<int> & ColOffsets() { return col_offsets; }
-   //! Read only access to the columns offsets for block starts
-   const Array<int> & ColOffsets() const { return col_offsets; }
-
-   /// Operator application
-   virtual void Mult (const Vector & x, Vector & y) const;
+extern template class TBlockOperator<const Operator>;
```

Comment:
Why are these `extern` statements needed, along with the explicit template instantations in `blockoperator.cpp`? Things compile fine without it as far as I can tell. Are you just trying to save compile time?

Reply:
It is explicitly instantiated in `blockoperator.cpp` so the compiler does not randomly places that instantiation when the `Block(Const)Operator` is used in the user code. Just a minor thing 😉

Reply:
I thought @hughcars wanted that ☝️ , but maybe not 😅 It keeps things compiling as they were before, i.e. BlockOperator in `blockoperator.cpp`/`.o` and does not bloat the user code/compilation. I think it is a good practice if you know what  argument the template will be used with most of the time. At other places, MFEM has usually a class for `double` and then a generalization with template for other things. However, it is a minor thing, I may remove it if you want...

Reply:
I suspect that without
```c++
extern template class TBlockOperator<const Operator>;
```
the compiler simply does not know that `TBlockOperator<const Operator>` was explicitly instantiated somewhere else, so it will basically never use that explicit instantiation. Note that the full template definition is available in the header, so the compiler will simply re-instantiate what it needs if it is not told that something was already explicitly instantiated somewhere else. That said, I'm not 100% sure this is exactly what's happening but this is what makes sense to me.